cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/SimpleRangeAnalysis.qll

```diff
@@ -0,0 +1,120 @@
+/**
+ * Wrapper for the semantic range analysis library that mimics the
+ * interface of the simple range analysis library.
+ */
+
+private import cpp
+private import semmle.code.cpp.ir.IR
+private import experimental.semmle.code.cpp.semantic.SemanticBound
+private import experimental.semmle.code.cpp.semantic.SemanticExprSpecific
+private import RangeAnalysis
+
+/**
+ * Gets the lower bound of the expression.
+ *
+ * Note: expressions in C/C++ are often implicitly or explicitly cast to a
+ * different result type. Such casts can cause the value of the expression
+ * to overflow or to be truncated. This predicate computes the lower bound
+ * of the expression without including the effect of the casts. To compute
+ * the lower bound of the expression after all the casts have been applied,
+ * call `lowerBound` like this:
+ *
+ *    `lowerBound(expr.getFullyConverted())`
+ */
+float lowerBound(Expr expr) {
+  exists(Instruction i, SemBound b | i.getAst() = expr and b instanceof SemZeroBound |
+    semBounded(getSemanticExpr(i), b, result, false, _)
+  )
```

Comment:
The QLDoc here was directly copied from the simple range analysis library. What I'm not convinced about is whether we satisfy the part of the QLDoc that talks about casts, or whether some additional logic is needed for that.

Reply:
I think the above implementation handles casts correctly. If `expr` is a conversion then `i.getAst() = expr` will be satisfied for some `ConvertInstruction`, and we'll compute the bound for the result of the conversion.

And if `expr` is not a conversion then `i.getAst() = expr` will be satisfied by some non-`ConvertInstruction`, and we'll compute the bound for the unconverted expression.

Now, I don't think we actually properly compute any useful bounds for `ConvertInstruction`s. But hopefully this will change soon 🤞.